dali/pipeline/operators/reader/loader/lmdb.h

```diff
@@ -24,71 +24,153 @@
+    CHECK_LMDB(mdb_dbi_open(mdb_transaction_, NULL, 0, &mdb_dbi_));
+    CHECK_LMDB(mdb_cursor_open(mdb_transaction_, mdb_dbi_, &mdb_cursor_));
+    MDB_stat stat;
+    CHECK_LMDB(mdb_stat(mdb_transaction_, mdb_dbi_, &stat));
+    mdb_size_ = stat.ms_entries;
+    printf("lmdb file %d %s has %ld entries\n", num_, path.c_str(), mdb_size_);
+    mdb_index_ = 0;
+    return true;
+  }
+  size_t GetSize() const { return mdb_size_; }
+  Index GetIndex() const { return mdb_index_; }
+  void SeekByIndex(Index index, MDB_val* key = 0, MDB_val* value = 0) {
+    MDB_val tmp_key, tmp_value;
+    if (0 == key) {
+      key = &tmp_key;
+    }
+    if (0 == value) {
+      value = &tmp_value;
+    }
+    //printf("index %ld, current_index %ld, size %ld\n", index, mdb_index_, mdb_size_);
+    if (index == 0) {
+      printf("lmdb %d %s rewind to the begin from %ld...\n", num_, db_path_.c_str(), mdb_index_);
+    }
+    DALI_ENFORCE(index >= 0 && index < mdb_size_);
+    if (index == 0) {
+      CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_FIRST));
+    } else if (index == mdb_size_ - 1) {
+      CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_LAST));
+    } else if (index == mdb_index_) {
+      CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_GET_CURRENT));
+    } else if (index == mdb_index_ - 1) {
+      CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_PREV));
+    } else if (index == mdb_index_ + 1) {
+      CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_NEXT));
+    } else if (index > mdb_index_) {
+      printf("lmdb exec a large step forward %s %ld->%ld\n", db_path_.c_str(),
+             mdb_index_, index);
+      for (Index i = mdb_index_; i < index; ++i) {
+        CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_NEXT));
+      }
     } else {
-      CHECK_LMDB(status, filename);
-      return true;
+      // index < mdb_index_
+      printf("lmdb exec a large step backward %s %ld->%ld\n", db_path_.c_str(),
+             mdb_index_, index);
+      for (Index i = index; i < mdb_index_; i++) {
+        CHECK_LMDB(mdb_cursor_get(mdb_cursor_, key, value, MDB_PREV));
+      }
     }
+    mdb_index_ = index;
   }
 
-  inline uint64_t LMDB_size(MDB_txn* txn, MDB_dbi dbi, const string &filename) {
-    MDB_stat stat;
-
-    CHECK_LMDB(mdb_stat(txn, dbi, &stat), filename);
-
-    return stat.ms_entries;
+  bool Close() {
+    if (mdb_cursor_) {
+      mdb_cursor_close(mdb_cursor_);
+      mdb_dbi_close(mdb_env_, mdb_dbi_);
+      mdb_cursor_ = nullptr;
+    }
+    if (mdb_transaction_) {
+      mdb_txn_abort(mdb_transaction_);
+      mdb_transaction_ = nullptr;
+    }
+    if (mdb_env_) {
+      mdb_env_close(mdb_env_);
+      mdb_env_ = nullptr;
+    }
+    return true;
   }
+};
 
-  inline void PrintLMDBStats(MDB_txn* txn, MDB_dbi dbi, const string &filename) {
-    MDB_stat stat;
-
-    CHECK_LMDB(mdb_stat(txn, dbi, &stat), filename);
+static int find_lower_bound(const std::vector<Index>& a, Index x) {
```

Comment:
can we use `std::lower_bound` instead?
http://www.cplusplus.com/reference/algorithm/lower_bound/

Reply:
Thank you. We can not use std::lower_bound. For example, `a = {0, 7, 16, 24}`, 
`find_lower_bound(a, 1)` return 0, which means the 1st entry is in lmdb0, 
while `std::lower_bound(a.begin(), a.end(), 1)-a.begin()` returns 1.